File: 91_Train/python/train.py
#!/usr/bin/env python3
# TRAIN
#
# Converted from BASIC to Python by Trevor Hobson

import random


def play_game():
    """Play one round of the game"""
    car_speed = random.randint(40, 65)
    time_difference = random.randint(5, 20)
    train_speed = random.randint(20, 39)
    print("\nA car travelling", car_speed, "MPH can make a certain trip in")
    print(time_difference, "hours less than a train travelling at", train_speed, "MPH")
    time_answer = 0
    while time_answer == 0:
        try:
            time_answer = float(input("How long does the trip take by car "))
        except ValueError:
            print("Please enter a number.")
    car_time = time_difference*train_speed/(car_speed-train_speed)
    error_percent = int(abs((car_time-time_answer)*100/time_answer)+.5)
    if error_percent > 5:
        print("Sorry. You were off by", error_percent, "percent.")
        print("Correct answer is", round(car_time, 6), "hours")
    else:
        print("Good! Answer within", error_percent, "percent.")


def main():
    print(" " * 33 + "TRAIN")
    print(" " * 15 + "CREATIVE COMPUTING  MORRISTOWN, NEW JERSEY\n")
    print("Time - speed distance exercise")

    keep_playing = True
    while keep_playing:
        play_game()
        keep_playing = input(
            "\nAnother problem (yes or no) ").lower().startswith("y")


if __name__ == "__main__":
    main()

File: 91_Train/lua/train.lua
print [[
            TRAIN
 CREATIVE COMPUTING  MORRISTOWN, NEW JERSY



 TIME - SPEED DISTANCE EXERCISE]]

math.randomseed(os.time())

local ERROR_MARGIN <const> = 5.0

function play()
    local car_speed = 25*math.random() + 40--Between 40 and 64
    local delta_time = 15*math.random() + 5--Between 5 and 19
    local train_speed = 19*math.random() + 20--Between 20 and 38

    print( string.format("\nA CAR TRAVELING AT %u MPH CAN MAKE A CERTAIN TRIP IN %u HOURS LESS THAN A TRAIN TRAVELING AT %u MPH.", car_speed, delta_time, train_speed) )

    local try = true
    local input
    while try do
        io.write("HOW LONG DOES THE TRIP TAKE BY CAR? ")
        input = io.read("n")
        if input == nil then
            print("<!>PLEASE INSERT A NUMBER<!>")
        else
            try = false
        end
        io.read()
    end

    local car_time = delta_time * train_speed / (car_speed - train_speed)
    local percent = ( math.abs(car_time-input) * 100 / car_time + .5)

    if percent > ERROR_MARGIN then
        print( string.format("SORRY. YOU WERE OFF BY %f PERCENT.", percent) )
    else
        print( string.format("GOOD! ANSWER WITHIN %f PERCENT.", percent) )
    end
    
    print( string.format("CORRECT ANSWER IS %f HOURS.", car_time) )
end

function game_loop()
    local keep_playing = true
    while keep_playing do
        play()
        io.write("\nANOTHER PROBLEM (YES OR NO)? ")
        answer = io.read("l")
        
        if not (answer == "YES" or answer == "Y" or answer == "yes" or answer == "y") then
            keep_playing = false
        end
    end

end

game_loop()